crates/noirc_evaluator/src/ssa_refactor/opt/simplify_cfg.rs

```diff
@@ -0,0 +1,190 @@
+//! This file contains the simplify cfg pass of the SSA IR.
+//!
+//! This is a rather simple pass that is expected to be cheap to perform. It:
+//! 1. Removes blocks with no predecessors
+//! 2. Inlines a block into its sole predecessor if that predecessor only has one successor.
+//! 3. Removes any block arguments for blocks with only a single predecessor.
+//! 4. Removes any blocks which have no instructions other than a single terminating jmp.
+//!
+//! Currently, only 2 and 3 are implemented.
+use std::collections::HashSet;
+
+use crate::ssa_refactor::{
+    ir::{
+        basic_block::BasicBlockId, cfg::ControlFlowGraph, function::Function,
+        instruction::TerminatorInstruction,
+    },
+    ssa_gen::Ssa,
+};
+
+impl Ssa {
+    /// Simplify each function's control flow graph by:
+    /// 1. Removing blocks with no predecessors
+    /// 2. Inlining a block into its sole predecessor if that predecessor only has one successor.
+    /// 3. Removing any block arguments for blocks with only a single predecessor.
+    /// 4. Removing any blocks which have no instructions other than a single terminating jmp.
+    ///
+    /// Currently, only 2 and 3 are implemented.
+    pub(crate) fn simplify_cfg(mut self) -> Self {
+        for function in self.functions.values_mut() {
+            simplify_function(function);
+        }
+        self
+    }
+}
+
+/// Simplify a function's cfg by going through each block to check for any simple blocks that can
+/// be inlined into their predecessor.
+fn simplify_function(function: &mut Function) {
+    let mut cfg = ControlFlowGraph::with_function(function);
+    let mut stack = vec![function.entry_block()];
+    let mut visited = HashSet::new();
+
+    while let Some(block) = stack.pop() {
+        if visited.insert(block) {
+            stack.extend(function.dfg[block].successors().filter(|block| !visited.contains(block)));
+        }
+
+        let mut predecessors = cfg.predecessors(block);
+
+        if predecessors.len() == 1 {
+            let predecessor = predecessors.next().expect("Already checked length of predecessors");
+            drop(predecessors);
+
+            // If the block has only 1 predecessor, we can safely remove its block parameters
+            remove_block_parameters(function, block, predecessor);
+
+            // Note: this function relies on `remove_block_parameters` being called first.
+            // Otherwise the inlined block will refer to parameters that no longer exist.
+            if try_inline_into_predecessor(function, &mut cfg, block, predecessor) {
+                continue;
```

Comment:
This `continue` seems inconsequential.

Reply:
My intent originally was that I expected more checks to be added here and if `try_inline_into_predecessor` succeeds and inlines the block, any later checks would be on a now-empty block so I wanted to prevent that from happening.
I've updated the code to remove the continue and add a comment to clarify the intent here.